### PR TITLE
Automated e2e test

### PR DESCRIPTION
**Pull request by hdbeukel**
Adds automated e2e tests for the main workflow. There are four test cases:
1. maize_v4 genome_wide
     - with DE (1 gene set)
     - network filtering (1 gene set)
     - 2nd gene annotation
     - no bps
2. maize_v4 locus_based
     - no DE
     - network filtering (1 gene set)
     - with bps
3. arabidopsis genome_wide
     - with DE (separate gene sets)
     - no network filtering
     - no 2nd gene annotation
     - bps
4. arabidopsis locus_based
     - no DE
     - no network filtering
     - no bps

### CODE REVIEW

**Comment by hdbeukel** [comment]
TODO before reviewing:
[ ] run on full example set (verify results manually)
[ ] try to setup automated GitHub action to run tests for pushes and PRs to `dev` and `main`

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+# ignore SLURM output and error files
+slurm.*.out
+slurm.*.err
+
+# ignore jupyter notebook checkpoints
+.ipynb_checkpoints/
+
+# python cache and compiled files
+__pycache__/
+*.pyc
```
Comment (nicomaper) [inline .gitignore:29]: Why do I see a negative sign in this?

File: bin/add_go_names.py
Comment (nicomaper) [inline bin/add_go_names.py:1]: I did not make this script, I think it was Dries, and later Thomas, but the change seems fine with me. Also, it makes sense because before you had to have an "ontologies" folder with the "go.obo" file but not anymore, right?

File: bin/getMotifCentricOutput_gw.py [hunk 2]
```diff
@@ -122,11 +125,11 @@ def parseArgs():
...
+    enr_stats = enr_stats.groupby(['dataset', 'motif', 'real_int', 'shuffled_int', 'p_val', 'enr_fold', 'adj_pval', 'pi_value', 'rank_pi_val']).agg({'gene_id': ','.join, 'family': lambda x: ','.join(sorted(set(x))), 'Any expressed gene': any}).reset_index().sort_values(by = 'rank_pi_val').drop('gene_id', axis = 1)
     
 if not expressed_genes_file:
 
-    enr_stats = enr_stats.groupby(['dataset', 'motif', 'real_int', 'shuffled_int', 'p_val', 'enr_fold', 'adj_pval', 'pi_value', 'rank_pi_val']).agg({'gene_id': ','.join, 'family': lambda x: ','.join(list(set(x)))}).reset_index().sort_values(by = 'rank_pi_val').drop('gene_id', axis = 1)
+    enr_stats = enr_stats.groupby(['dataset', 'motif', 'real_int', 'shuffled_int', 'p_val', 'enr_fold', 'adj_pval', 'pi_value', 'rank_pi_val']).agg({'gene_id': ','.join, 'family': lambda x: ','.join(sorted(set(x)))}).reset_index().sort_values(by = 'rank_pi_val').drop('gene_id', axis = 1)
```
Comment (nicomaper) [inline bin/getMotifCentricOutput_gw.py:132]: I was puzzled by this change because I didn't know applying "sorted" to a set would convert it to a list, but is seems correct to me!

File: test.sh
[the remark this answers is not in the document]
Comment (nicomaper) [reply test.sh:1]: Is this not redundant with what's in the yml? Should this be pushed or be put in .gitignore?

File: workflows/miniac_gw.nf [hunk 3]
```diff
@@ -191,14 +198,11 @@ process GOenrichment{
 
         grep -v '#' ${acr_file_name}_GO_enrichment_rawOut.txt > ${acr_file_name}_GO_enrichment_noheader.txt
 
-        mkdir ontologies
-        mv $obo_file ontologies
```
Comment (nicomaper) [inline workflows/miniac_gw.nf:195]: Nice! Now I see the effect of the change in the ontologies script :)

Comment (nicomaper) [reply workflows/miniac_gw.nf:195]: This was not a review but a comment sorry

**Review by nicomaper — CHANGES_REQUESTED** [review]
Wow this test implementation is so nice, Herman, thanks a lot for investing time in it! I'm very impressed also by in how little time you managed to make it happen (probably for a computer scientist is normal, but for me, it would've taken much more time).

Many of the "reviews" are not actually reviews, but comments. There are a few things where I am confused about which files should be pushed and which files should not. Perhaps it is easier that you explain to me in person. Also, there is the comment about the commented code, and also whether that piece of code should be removed or not (giving you your own medicine haha :))

**Comment by hdbeukel** [comment]
Wow @nicomaper thanks for the detailed review! I have skimmed through your comments and many make a lot of sense. I will fist answer your comments that are actually questions and then make some changes. I'll let you know when I am done.

File: .gitignore (hunk 1 above)
Comment (hdbeukel) [reply .gitignore:29]: Because the .gitignore file has no newline (`\n`) on the final line. GitHub flags this as a potential mistake, since the Unix convention is that every line should end with a newline character, also the last line in your file. See https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline. You may wonder why, but it makes sense. For example if you process text files with regular expressions then you want to be sure it works on all lines. If the final line is different in the sense that there is no trailing newline, you might make mistakes.

Long story short: I will add a newline at the end of the file.

Comment (hdbeukel) [reply .gitignore:29]: FYI: if you edit text files with command line editors such as `nano` or `vi`, they will automatically add a newline character at the end of the last line if it is missing.

File: bin/add_go_names.py
Comment (hdbeukel) [reply bin/add_go_names.py:1]: Yes indeed. I got in trouble here with the test runs as these are executed in a subfolder. After analysing the code, I realised it was doing redundant stuff. The ontology file was being moved to another folder and then the relative path to this new folder was constructed in a weird way from within the Python script. When calling the script, you already have access to the original path to the .obo file, so I decided to simply pass this as input to the Python script. This reduced complexity and fixed the test flow.

File: bin/getMotifCentricOutput_gw.py (hunk 2 above)
Comment (hdbeukel) [reply bin/getMotifCentricOutput_gw.py:132]: Yes indeed: https://www.programiz.com/python-programming/methods/built-in/sorted

File: tests/data/zma_v4/zma_v4_noncod_merged_chr1.bed
[the remark this answers is not in the document]
Comment (hdbeukel) [reply tests/data/zma_v4/zma_v4_noncod_merged_chr1.bed:1]: I went for the same approach you used for the real data files: everything _except_ the motif mapping files is pushed into the repository (because these files are relatively small) but the motif mapping files are not (because they are huge). Even after filtering the maize test data for chromosome 1 only, the genome-wide and locus-based motif mapping files together add up to about 500MB. Arabidopsis files have similar size, since here we didn't apply any filtering.

The `zma_v4_noncod_merged_chr1.bed` file you referred to, is not a motif mapping file, but the non-coding genome space (which is a small file).

So indeed the motif mapping files are downloaded each time before the tests can run.

File: tests/data/ath/.keep
[the remark this answers is not in the document]
Comment (hdbeukel) [reply tests/data/ath/.keep:1]: I removed the whole folder `tests/data/ath` since for Arabidopsis we are testing on the full genome, so this is the same `data` for actual runs. For maize I renamed `tests/data/zma_v4` to `tests/data/zma_v4_chr1` to make it more clear that this is specific reduced data for tests only, which is why it is stored in the `tests` folder.

File: tests/mini_ac.nf.test.snap
[the remark this answers is not in the document]
Comment (hdbeukel) [reply tests/mini_ac.nf.test.snap:1]: This is indeed automatically generated by `nf-test`. The first time you run the tests, it creates the snapshot because we used `snapshot(...).match()` in the test asserts. For subsequent runs it then expects that the output will be the same. For output _files_ the snapshot contains a hash of the output file, not the entire output, and then `nf-test` will check that the hash remains the same in subsequent runs.

File: test.sh
[the remark this answers is not in the document]
Comment (hdbeukel) [reply test.sh:1]: As discussed, I removed this script since we will now let GitHub actions do the heavy lifting instead of running the tests on the cluster manually.

File: nf-test.config
[the remark this answers is not in the document]
Comment (hdbeukel) [reply nf-test.config:1]: No this is supposed to be in the root folder. This is the `nf-test` config, from which the `nf-test` command know e.g. where the tests are to be found (in the `tests` directory). The other config file within the `tests` directory is the actual MINI-AC config that will be used when running the workflow tests.

File: tests/nextflow.config
[the remark this answers is not in the document]
Comment (hdbeukel) [reply tests/nextflow.config:1]: As mentioned above I have removed the SLURM-specific configuration and reduced this to what is needed for GitHub actions.

**Review by nicomaper — APPROVED** [review]
